pkg/hooks/manager.go

```diff
@@ -252,83 +257,169 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
 			return nil, err
 		}
 		domainSpecXML = result.GetDomainXML()
+	case hooksV1alpha3.Version:
+		client := hooksV1alpha3.NewCallbacksClient(conn)
+		result, err := client.OnDefineDomain(ctx, &hooksV1alpha3.OnDefineDomainParams{
+			DomainXML: domainSpecXML,
+			Vmi:       vmiJSON,
+		})
+		if err != nil {
+			log.Log.Reason(err).Error("Failed to call OnDefineDomain")
+			return nil, err
+		}
+		domainSpecXML = result.GetDomainXML()
 	default:
 		log.Log.Errorf("Unsupported callback version: %s", callback.Version)
 	}
 
 	return domainSpecXML, nil
 }
 
+func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cloudinit.CloudInitData) ([3][]byte, error) {
+	empty := [3][]byte{}
+	vmiJSON, err := json.Marshal(vmi)
+	if err != nil {
+		return empty, fmt.Errorf("failed to marshal VMI spec, err: %v", err)
+	}
+
+	// To be backward compatible to sidecar hooks still expecting to receive the cloudinit data as a
+	// CloudInitNoCloudSource object,
+	// we need to construct a CloudInitNoCloudSource object with the user- and networkdata from the
+	// cloudInitData object.
+	cloudInitNoCloudSource := v1.CloudInitNoCloudSource{
+		UserData:    cloudInitData.UserData,
+		NetworkData: cloudInitData.NetworkData,
+	}
+	cloudInitNoCloudSourceJSON, err := json.Marshal(cloudInitNoCloudSource)
+	if err != nil {
+		return empty, fmt.Errorf("failed to marshal CloudInitNoCloudSource: %v, err: %v", cloudInitNoCloudSource, err)
+	}
+
+	cloudInitDataJSON, err := json.Marshal(cloudInitData)
+	if err != nil {
+		return empty, fmt.Errorf("failed to marshal CloudInitData, err: %v", err)
+	}
+
+	return [3][]byte{cloudInitDataJSON, cloudInitNoCloudSourceJSON, vmiJSON}, nil
+}
+
+func preCloudInitIsoValidateResult(dataSource cloudinit.DataSourceType, initData, noCloudSource []byte) (*cloudinit.CloudInitData, error) {
+	var resultData *cloudinit.CloudInitData
+	err := json.Unmarshal(initData, &resultData)
+	if err != nil {
```

Comment:
nit: Inline check

Reply:
I'll fix the nits in the next iteration.